homeassistant/components/nextbus/__init__.py

```diff
@@ -13,15 +13,19 @@
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up platforms for NextBus."""
     entry_agency = entry.data[CONF_AGENCY]
+    entry_stop = entry.data[CONF_STOP]
+    coordinator_key = f"{entry_agency}-{entry_stop}"
 
-    coordinator: NextBusDataUpdateCoordinator = hass.data.setdefault(DOMAIN, {}).get(
-        entry_agency
+    coordinator: NextBusDataUpdateCoordinator | None = hass.data.setdefault(
+        DOMAIN, {}
+    ).get(
+        coordinator_key,
     )
     if coordinator is None:
         coordinator = NextBusDataUpdateCoordinator(hass, entry_agency)
-        hass.data[DOMAIN][entry_agency] = coordinator
+        hass.data[DOMAIN][coordinator_key] = coordinator
 
-    coordinator.add_stop_route(entry.data[CONF_STOP], entry.data[CONF_ROUTE])
+    coordinator.add_stop_route(entry_stop, entry.data[CONF_ROUTE])
```

Comment:
Can we maybe add tests for something I want to verify? Once you create a coordinator in a config entry context, it will bind the coordinator to the config entry and it will shutdown the coordinator when the config entry unloads 

https://github.com/home-assistant/core/blob/f006716173eb16848b0bcf53d8309fcdea803019/homeassistant/helpers/update_coordinator.py#L123-L124

So I think this setup might not even work here and cause the other config entries to stop working when for example a config entry is deleted

Reply:
Sorry. I'm not following.

On setup the coordinator is being added to the domain under a key that is possibly shared between different entities.

In the remove config function, the relevant coordinator is looked up, the entity specific data is popped out, then there is a check to see if there are remaining routes the coordinator is still responsible for, only if this is empty is it deleted.

What would a test like this look like?

Reply:
But check the code that I linked, the config entry that created the coordinator would shut down the coordinator while the entry got unloaded.

Set up 2 entries which share that coordinator, then unload one, skip time and check if the coordinator still works like expected

Reply:
Oh, hmm. I've been using this pattern since the original creation of this component. I figured this was `async_unload_entry` was supposed to handle, so I managed the logic of when to stop listening there. It looks like this `async_on_unload` was added only last year and may have introduced a bug.

It looks like when implemented it was suggested to make it "opt in" to auto cleanup, but instead no optionally was provided, with the intent of making sure this "just works" for typical use cases. This could be resolved by adding an "opt out" mechanism which would preserve the "just works" aspect of "fixing" zombie coordinators.

I've managed another way to do it by overriding the unload function and checking there.

I've got a test ready but trying to figure out how to "skip time". Any pointers?

Reply:
Figured it out. I was about to confirm the issue you mentioned with a test and then verify my patch resolves it.

Reply:
Use the freezer! Im on mobile so I can't point to code right now. But a quick search will yield results.

I think it would be nice if we can go for a separate setup (so every config entry it's own coordinator). This way we know everything works like expected and then we make sure we have tests. And then we can always discover ways to optimise this. There are other classes (like the Base coordinator protocol) which don't have that limitation, so maybe that works for you. But I think it would be nice to get it to a working state before we start to optimise

Reply:
Well, the efficiency and the working state fix are one in the same here. There is an undocumented API that is being used and it appears that too many requests and it starts to 404. This patch reduces reuse of the coordinator where it doesn't add efficiency as well as allows reuse when it will. This should reduce 404s, but also prevent a 404 from killing updates from unrelated entities.

Reply:
Haha okay, I was scared when you said "I've managed another way to do it by overriding the unload function and checking there.". I thought you did some crazy thing to avoid the async_on_unload, hence my comment. 

I'll review later

Reply:
Oh. Not too crazy. It’s a public method, so I added a check and then call super.